### ISSUE TITLE
.ipynb debug cell -- debugger fails to hit breakpoint

### ISSUE THREAD

**Issue by p-i-** [microsoft/vscode-jupyter#15914]
Type: <b>Bug</b>

# TLDR (Edited 9 Aug)
In my case the problem was fixed by toggling "Jupyter: Debug Just My Code" in Settings to ON.

# Original message ...

I have VSCode opened in $root. I have $root/.venv/... (Python 3.12.3, freshly installed using pyenv (macos))

I create test.ipynb

First cell: `a = 1`, Second cell: `b = 2`. I execute the first cell. ok. I set a breakpoint on the first (and only) line of the second cell and debug-execute it.

There's no visual indication the breakpoint has been hit. And attempting to step-{in/out/over} lands me in alien code (see screenshot).

Now I try to create a fresh test case. Quit VSCode, create test/ folder, create test/.venv/ (from pyenv, identically as done per fail-case), open VSCode in test/, create x.ipynb, 2 cells `a = 2` and `b = 2`, execute the first. ok. set breakpoint on second and debug-execute the second.

It hits!  It works!

So something in my $root project is mashing the debugger.

This is bizarre and frustrating. It means I'm going to have to copy my entire codebase into a fresh project just to get the debugger working in .ipynb files.

Extension version: 2024.7.0
VS Code version: Code 1.92.0 (b1c0a14de1414fcdaa400695b4db1c0799bc3124, 2024-07-31T23:26:45.634Z)
OS version: Darwin arm64 23.5.0
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Apple M2 (8 x 2400)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled<br>webnn: disabled_off|
|Load (avg)|3, 3, 2|
|Memory (System)|24.00GB (2.33GB free)|
|Process Argv|--crash-reporter-id f10d97cd-2115-4dba-a34a-07be9312995a|
|Screen Reader|no|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vspor879:30202332
vspor708:30202333
vspor363:30204092
vstes627:30244334
vscorecescf:30445987
vscod805cf:30301675
binariesv615:30325510
vsaa593cf:30376535
py29gd2263:31024239
vscaac:30438847
c4g48928:30535728
azure-dev_surveyone:30548225
2i9eh265:30646982
962ge761:30959799
pythongtdpath:30769146
welcomedialog:30910333
pythonnoceb:30805159
asynctok:30898717
pythonregdiag2:30936856
pythonmypyd1:30879173
2e7ec940:31000449
pythontbext0:30879054
accentitlementst:30995554
dsvsc016:30899300
dsvsc017:30899301
dsvsc018:30899302
cppperfnew:31000557
dsvsc020:30976470
pythonait:31006305
dsvsc021:30996838
jg8ic977:31013176
pythoncenvpt:31062603
a69g1124:31058053
dvdeprecation:31068756
dwnewjupytercf:31046870
impr_priority:31102340
refactort:31108082
ccplc:31103425
pythonrstrctxt:31103193
wkspc-onlycs-c:31106320
wkspc-ranged-c:31107834

```

</details>

**Comment by p-i-** [comment]
Test case (works):

<img width="930" alt="Screenshot 2024-08-02 at 14 30 32" src="https://github.com/user-attachments/assets/a453f5f5-f49e-4e60-8dd7-5be7263caec0">


Fail case (my project):
<img width="935" alt="Screenshot 2024-08-02 at 14 31 21" src="https://github.com/user-attachments/assets/7f1f5d79-826f-4581-8198-7c5d0977d023">

If I click step {over/out/in} on the fail-case:
<img width="616" alt="Screenshot 2024-08-02 at 14 37 38" src="https://github.com/user-attachments/assets/f77e4447-ca88-4752-a949-fb8065f14e5e">

**Comment by p-i-** [comment]
<img width="272" alt="Screenshot 2024-08-02 at 14 43 52" src="https://github.com/user-attachments/assets/5a587d5d-ceb8-47df-9ff4-e1c76f2c9bb8">

^ That's how I'm invoking debug on a cell

**Comment by p-i-** [comment]
Here are the TRACE logs for Output -> Jupyter:

FAIL:
[fail.txt](https://github.com/user-attachments/files/16471584/fail.txt)

WIN:
[win.txt](https://github.com/user-attachments/files/16471587/win.txt)

**Comment by p-i-** [comment]
As a follow-up, this has hit me again.

To resolve it, I created a fresh folder, created a ./.venv, opened VSCode, copied my .ipynb and .py file into this new folder, and was able to successfully use the debugger in my .ipynb.

- Setting breakpoints either in the .ipynb or the .py and debug-running the cell, they would hit.
- Checking 'raised exceptions' in the "Breakpoints" pane, execution would indeed suspend at the exception, allowing me to use "Debug Console" pane to inspect.

Then I left VSCode running and went to sleep. And now first thing in the morning, it is again failing.

I haven't done anything fancy. All I've been doing is tweaking a numpy-heavy .py file and an .ipynb that uses it. I haven't done anything else on my mac, other than watch a YouTube video in a Chrome tab.

In fact my "fresh folder" was the folder I used for my working testcase (first screenshot).

And re-executing the test now fails. On exactly the same x.ipynb file!

I've looked through all available logs in Output pane, and see no anomaly.

This is a serious problem :/

**Comment by p-i-** [comment]
As per https://stackoverflow.com/questions/56794940/vscode-why-isnt-debugger-stopping-at-breakpoints toggling "Jupyter: Debug Just My Code" in Settings to ON has gotten it working.

<img width="1235" alt="Screenshot 2024-08-04 at 10 42 57" src="https://github.com/user-attachments/assets/a0729265-6979-438c-8fdc-8cf3ee2446cf">

Why can't I see the setting I just changed in `~/Library/Application\ Support/Code/User/settings.json`? Even after a VSCode restart it's not there. And I have no .vscode/ folder in my project. I thought every settings-change creates a new entry at the bottom of this JSON list.

**Comment by DonJayamanne** [comment]
Than you for filling this issues and sorry you are running into this
And Thank you for your detailed analysis and screenshot

I think the final conclusion would that the settings aren’t getting saved/applied correctly.

Is that correct?

**Comment by p-i-** [comment]
Let me clarify my last post. There's two complete separate things in it.

1. Toggling "Jupyter: Debug Just My Code" to ON fixes the "debugger not working correctly" issue. Now this is a bizarre fix. So the issue should be considered as a bug IMO.

2. After toggling this setting, I decided to inspect my User/Settings.json so that I could copy down the exact key for this setting, and was surprised to find that the setting isn't getting written to that file. I doubt this is connected to the original issue. But I am curious as to why I can't find this setting. I just checked again today and it is still not there.

**Comment by DonJayamanne** [comment]
> as to why I can't find this setting. I just checked again today and it is still not there.

Default values aren't written into settings.json
By default this value is turn on, hence you may have disabled this at some point in time.


![Image](https://github.com/user-attachments/assets/85ee2ea1-881d-4aba-ae85-4db64a8860b4)


> I doubt this is connected to the original issue. 

Given that you are not debugging any 3rd party code and just a simple cell and its not hitting your break point.
I agree, that the setting should not really make a difference, however for now I would suggest enabling the value as that is the default value.

**Comment by p-i-** [comment]
Please someone remove the info-needed tag.

**Comment by edervishaj** [comment]
I have the same problem, the debugger does not stop. Moreover, after I stop it, the little arrow that allows me to select "Debug Cell" disappears.

**Comment by p-i-** [comment]
I've also bumped into that disappearing debug-cell option.

As I remember, closing and restarting VSCode shakes it loose.

I haven't been able to isolate a clean consistent minimal fail, so I haven't reported that oddity yet.

There are still quite a few quirks with running .ipynbs in VSCode. I wish they had someone on the team that uses .ipynbs a lot -- they would surely bump into these frustration points.

Or even a Discord where if some engineer hits such a problem, a core-dev could remote-inspect their system over screenlink. I'd be more than happy to allow a core-dev to dig around in my box if it would lead to a seamless UX.

**Comment by edervishaj** [comment]
This is not the only issue I face. After stopping the debugger (because it does not stop) and relaunching it, I get the error "Couldn't find a debug adapter for debug type python". This started occurring after the last update of VSCode/Python extension. I tried downgrading everything, the problem persists.

**Comment by p-i-** [comment]
@edervishaj Did you try deleting and re-creating your Python .venv? It might be possible that it is corrupted somehow.

**Comment by edervishaj** [comment]
> @edervishaj Did you try deleting and re-creating your Python .venv? It might be possible that it is corrupted somehow.

I created a new environment altogether, did not help at all. The debugger stops but after the first step over it throws the above error.

**Comment by DonJayamanne** [comment]
> After toggling this setting, I decided to inspect my User/Settings.json so that I could copy down the exact key for this setting, and was surprised to find that the setting isn't getting written to that file. I doubt this is connected to the original issue. But I am curious as to why I can't find this setting. I just checked again today and it is still not there.

@p-i- Please can you create a separate issue for this, then we can close this. 
Also, when creating a new issue, please can you create the issue in the VS Code repo, as thats un-related to Jupyter extension (its a functionality provided by VS Code itself)

> This is not the only issue I face. After stopping the debugger (because it does not stop) and relaunching it, I get the error "Couldn't find a debug adapter for debug type python". This started occurring after the last update of VSCode/Python extension. I tried downgrading everything, the problem persists.

@edervishaj Please can you create a new issue and please do share the following logs in that new issue
Please could you enable logging as follows:
* Open settings & go into `Jupyter->Logging`
* Change the value to `trace`
* Reload VS Code,
* Attempt to repro the issue & then copy the output from the `Jupyter` output panel (use the command `Jupyter: Show Output` to get to the logs).
